vividus-extension-selenium/src/main/java/org/vividus/bdd/steps/ui/BarcodeSteps.java

```diff
@@ -67,15 +74,18 @@ public BarcodeSteps(ScreenshotTaker screenshotTaker, BarcodeActions barcodeActio
     @When("I scan barcode from screen and save result to $scopes variable `$variableName`")
     public void scanBarcode(Set<VariableScope> scopes, String variableName) throws IOException
     {
-        BufferedImage screenshotPath = screenshotTaker.takeViewportScreenshot();
+        BufferedImage viewportScreenshot = screenshotTaker.takeViewportScreenshot();
         try
         {
-            String result = barcodeActions.scanBarcode(screenshotPath);
+            String result = barcodeActions.scanBarcode(viewportScreenshot);
             bddVariableContext.putVariable(scopes, variableName, result);
         }
         catch (NotFoundException e)
         {
             softAssert.recordFailedAssertion("There is no barcode on the screen", e);
+            Attachment attachment = new Attachment(ImageTool.toByteArray(viewportScreenshot), "Viewport Screenshot",
```

Comment:
Why `@TakeScreenshotOnFailure` annotating not enough?

Reply:
1) performance: duplicate operation taking one more screenshot
2) timing: the viewport screenshot contains the most relevant application state, if we take screenshot some time later, the state of the application may be changed (e.g. animations)